va/caa.go

```diff
@@ -317,24 +317,54 @@ func checkAccountURI(accountURI string, accountURIPrefixes []string, accountID i
 	return false
 }
 
-// Given a CAA record, assume that the Value is in the issue/issuewild format,
-// that is, a domain name with zero or more additional key-value parameters.
-// Returns the domain name, which may be "" (unsatisfiable), and a tag-value map of parameters.
+// extractIssuerDomainAndParameters extracts the domain and parameters (if any)
+// from a issue/issuewild CAA record. This follows sections 5.2 and 5.3 of the
+// RFC 6844bis draft (https://tools.ietf.org/html/draft-ietf-lamps-rfc6844bis-00),
+// where all components are semi-colon separated. The domain name (which may be
+// an empty string in the unsatisfiable case) and a tag-value map of parameters
+// are returned.
 func extractIssuerDomainAndParameters(caa *dns.CAA) (domain string, parameters map[string]string) {
 	isIssueSpace := func(r rune) bool {
 		return r == '\t' || r == ' '
 	}
 
-	v := strings.SplitN(caa.Value, ";", 2)
-	domain = strings.TrimFunc(v[0], isIssueSpace)
+	// Semi-colons (ASCII 0x3B) are prohibited from being specified in the
+	// parameter tag or value, hence we can simply split on semi-colons.
+	parts := strings.Split(caa.Value, ";")
+	domain = strings.TrimFunc(parts[0], isIssueSpace)
+
 	parameters = make(map[string]string)
+	for _, parameter := range parts[1:] {
+		valid := true
+
+		// A parameter tag cannot include equal signs (ASCII 0x3D),
+		// however they are permitted in the value.
+		tv := strings.SplitN(parameter, "=", 2)
+		if len(tv) != 2 {
+			// TODO(4a6f656c): Should we be silently ignoring incorrectly
+			// formatted parameters?
+			continue
+		}
+
+		tag := strings.TrimFunc(tv[0], isIssueSpace)
+		for _, r := range []rune(tag) {
+			// ASCII alpha/digits.
+			if r < 0x30 || r > 0x39 && r < 0x41 || r > 0x5a && r < 0x61 || r > 0x7a {
```

Comment:
This can be done somewhat more readably by using the actual ASCII runes, i.e. `('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || ('0' <= r && r <= '9')`.

Reply:
See below - if we change the value validation, then this is easily changed, otherwise I think having consistency is preferable.